### DISCUSSION TITLE
Is there available a macro with package name during processing `%files` section?

### DISCUSSION THREAD

**Discussion by voxik** [rpm-software-management/rpm#2876] [Q&A]
In `ruby` package, there are sections such as:

~~~
%files -n rubygem-psych
%{ruby_libdir}/psych
%{ruby_libdir}/psych.rb
%{ruby_libarchdir}/psych.so
%gem_extdir_mri psych}
%dir %{gem_instdir psych}
%{gem_libdir psych}
%{gem_spec -d psych}
~~~

And as it can be seen, the `psych` argument is there pretty repetitive. So I wonder, is there by a chance macro, which would contain name of the `%files` section which is actually processed?

Yes, there is the `rubygem-` prefix, but ...

Or maybe I could set some `files_name` variable on top of each `%files` section?

**Comment by pmatilai** [comment]
If you have lots of those with nothing but the name changing, no macros will help with that. Generate those files manifests programmatically from %install and include with %files -f. Or use dynamic spec generation to generate even those.

**Comment by voxik** [comment]
> If you have lots of those with nothing but the name changing, no macros will help with that.

Since e.g. the `gem_libdir` is macro, it could instead of accepting the parameter use some macro on background. So the question is can we have such macro?

> Generate those files manifests programmatically from %install and include with %files -f. Or use dynamic spec generation to generate even those.

For me, the `%files` section is not about getting some random files into package. If I greatly simplify the problem, the files section such as:

~~~
%files
%{_libdir}
~~~

is not the way to go and generated file list are very similar to this. While they allow to include everything what is on file system, they don't cover my back and they don't tell mi if e.g. soname has changed, if there are new files, which need review or if some files gone missing.

But maybe I just like doing things the hard way 🤷

**Reply by pmatilai** [reply]
No no. What I'm telling you is to script the file section generation. Something like

```
for n in psych mental; do
cat << EOF > ${n}.txt
%{ruby_libdir}/${n}
%{ruby_libdir}/${n}.rb
%{ruby_libarchdir}/${n}.so
#...
EOF
done
```

...and include from %files with -f. Or make that a spec local macro that takes the package name and produces the content inline.

**Reply by voxik** [reply]
I think my example was not clear enough. Lets concentrate on the later part:

~~~
%gem_extdir_mri psych}
%dir %{gem_instdir psych}
%{gem_libdir psych}
%{gem_spec -d psych}
~~~

If there was some macro with package name, I think I could simplify this:

~~~
%{gem_extdir_mri}
%dir %{gem_instdir}
%{gem_libdir}
%{gem_spec -d}
~~~

I'd probably left the part at the beginning untouched.

But still, seeing where you go and admitting it has some merit, I'd probably go without the cycle:

~~~
n=psych
cat << EOF > ${n}.txt
%{ruby_libdir}/${n}
%{ruby_libdir}/${n}.rb
%{ruby_libarchdir}/${n}.so
#...
EOF

n=mental
cat << EOF > ${n}.txt
%{ruby_libdir}/${n}
%{ruby_libdir}/${n}.rb
#...
EOF
~~~

Maybe because the `mental` does not have the .so file.

But then having the `%files` section feels useless. I think there was similar discussion about the `%excludes`.

**Reply by voxik** [reply]
But maybe the `%files` section should be obsolete and replaced by file lists ...

**Comment by pmatilai** [comment]
> If there was some macro with package name, I think I could simplify this:
> 
> %{gem_extdir_mri}
> %dir %{gem_instdir}
> %{gem_libdir}
> %{gem_spec -d}

If you declared these macros in terms including some %{name} then ... you'd have X number of such identical %files sections, now just even more redundant. 

I really don't understand what you're trying to achieve.

**Reply by voxik** [reply]
Please see the last several files section of [ruby.spec](https://src.fedoraproject.org/rpms/ruby/blob/rawhide/f/ruby.spec#_1563-1649). Granted, there is a lot of similarities between the `%files` sections, but they are not the same.

On top of that, some of them are also similar to the dual life packages, e.g. [rubygem-minitest](https://src.fedoraproject.org/rpms/rubygem-minitest/blob/rawhide/f/rubygem-minitest.spec). It would be neat if the dual life packages could have the same `%files` sections.

**Reply by pmatilai** [reply]
Yes, I get that you have repetitive sections. Like I've been telling you: script it (build scriptlets are shell scripts, you can have functions in there) or macroize it, with parameters to allow for the variety. The package name in macros would not be helpful here at all.
